src/Workspaces/Core/Portable/Options/Option`1.cs

```diff
@@ -70,33 +50,45 @@ public Option(string feature, string name, T defaultValue, params OptionStorageL
         }
 
         internal Option(string feature, OptionGroup group, string name, T defaultValue, params OptionStorageLocation[] storageLocations)
+            : this(feature, group, name, defaultValue, storageLocations.ToImmutableArray())
         {
-            if (string.IsNullOrWhiteSpace(feature))
-            {
-                throw new ArgumentNullException(nameof(feature));
-            }
+        }
 
-            if (string.IsNullOrWhiteSpace(name))
-            {
-                throw new ArgumentException(nameof(name));
-            }
+        internal Option(string feature, OptionGroup group, string name, T defaultValue, ImmutableArray<OptionStorageLocation> storageLocations)
+            : this(new OptionDefinition(feature, group, name, defaultValue, typeof(T), isPerLanguage: false), storageLocations)
+        {
+        }
 
-            this.Feature = feature;
-            this.Group = group ?? throw new ArgumentNullException(nameof(group));
-            this.Name = name;
-            this.DefaultValue = defaultValue;
-            this.StorageLocations = storageLocations.ToImmutableArray();
+        internal Option(OptionDefinition optionDefinition, ImmutableArray<OptionStorageLocation> storageLocations)
+        {
+            _optionDefinition = optionDefinition;
+            this.StorageLocations = storageLocations;
         }
 
+        OptionGroup IOptionWithGroup.Group => this.Group;
+
         object? IOption.DefaultValue => this.DefaultValue;
 
         bool IOption.IsPerLanguage => false;
 
-        OptionGroup IOptionWithGroup.Group => this.Group;
+        OptionDefinition IOption2.OptionDefinition => _optionDefinition;
+
+        bool IOption2.Equals(IOption2? other) => Equals(other);
 
-        public override string ToString()
+        public override string ToString() => _optionDefinition.ToString();
+
+        public override int GetHashCode() => _optionDefinition.GetHashCode();
+
+        public override bool Equals(object? obj) => Equals(obj as IOption2);
```

Comment:
IOption2 and not IOption? Is there some strange equality that might happen here, or non-reflexive equality? (I wouldn't have imagined this type implemented IOption2 at all...)

Reply:
To ensure that all our option sub-types are equatable, I introduced `bool Equals(object? obj)` and `bool Equals(IOption2? obj)` in `IOption2` interface (I cannot add to `IOption` as it would be a breaking change). Sub-types needs to implement these interface methods in terms of `IOption2` anyways for the second method, so I shared implementation using `IOption2` parameter.

https://github.com/dotnet/roslyn/pull/42323/files#diff-a0afe879ce7357f9a56acc378f5c7ec6R19-R23

```csharp
internal interface IOption2 : IOption
{
       ...

        // Ensure that all the sub-types are equatable
        // by requiring them to provide an implementation for Equals/GetHashCode.
        bool Equals(IOption2? other);
        bool Equals(object? other);
        int GetHashCode();

}
```

Reply:
Do we have some tests for this? This seems really easy to break.

Reply:
I actually removed these interface method based on @sharwell's suggestion. We now have unit tests to verify that options are indeed equatable.